File: docs/checker-widget/src/engine.ts
/**
 * TypeScript port of packages/core/src/opencomplai_core/compliance_checker/engine.py
 * Deterministic FLI-parity compliance checker — checker version fli-2025-07-28.
 *
 * Must produce identical output to the Python evaluate() for all 17 golden
 * fixtures in packages/core/tests/fixtures/fli_golden/.
 */
import {
  getObligation,
  getStatusChange,
  ObligationItem,
  StatusChangeItem,
} from "./catalog";

export const CHECKER_VERSION = "fli-2025-07-28";

export type EntityType =
  | "provider"
  | "deployer"
  | "distributor"
  | "importer"
  | "product_manufacturer"
  | "authorised_rep";

export interface CheckerResult {
  checker_version: string;
  in_scope: boolean;
  is_high_risk: boolean;
  is_prohibited: boolean;
  effective_entity: EntityType | null;
  status_changes: StatusChangeItem[];
  obligations: ObligationItem[];
  determination_path: string[];
  answers: Record<string, unknown>;
  session_id: string | null;
}

// ── helpers ──────────────────────────────────────────────────────────────────

function answerBool(
  answers: Record<string, unknown>,
  key: string,
  defaultVal = false
): boolean {
  const v = answers[key];
  if (v === undefined || v === null) return defaultVal;
  if (typeof v === "boolean") return v;
  if (typeof v === "string") return ["true", "yes", "1"].includes(v.toLowerCase());
  return Boolean(v);
}

function answerEntity(answers: Record<string, unknown>): EntityType {
  const raw = answers["e1_entity_type"];
  if (typeof raw === "string") return raw as EntityType;
  return "provider";
}

function determineHighRisk(answers: Record<string, unknown>): boolean {
  const hr1 = answerBool(answers, "hr1_annex_i");
  const hr2 = answerBool(answers, "hr2_annex_iii");
  if (!hr1 && !hr2) return false;
  if (answerBool(answers, "hr3_art_6_3")) return false;
  if (answerBool(answers, "hr4_narrow_task")) return false;
  if (answerBool(answers, "hr5_no_significant_risk")) return false;
  if (answerBool(answers, "hr6_accessory")) return false;
  return true;
}

function dedupeIds(ids: string[]): string[] {
  const seen = new Set<string>();
  return ids.filter((id) => {
    if (seen.has(id)) return false;
    seen.add(id);
    return true;
  });
}

function dedupeObligations(items: ObligationItem[]): ObligationItem[] {
  const seen = new Set<string>();
  return items.filter((item) => {
    if (seen.has(item.id)) return false;
    seen.add(item.id);
    return true;
  });
}

function dedupeStatusChanges(items: StatusChangeItem[]): StatusChangeItem[] {
  const seen = new Set<string>();
  return items.filter((item) => {
    if (seen.has(item.id)) return false;
    seen.add(item.id);
    return true;
  });
}

function buildResult(opts: {
  in_scope: boolean;
  path: string[];
  statusIds: string[];
  obligationIds: string[];
  is_high_risk?: boolean;
  is_prohibited?: boolean;
  effective_entity?: EntityType | null;
  answers: Record<string, unknown>;
}): CheckerResult {
  return {
    checker_version: CHECKER_VERSION,
    in_scope: opts.in_scope,
    is_high_risk: opts.is_high_risk ?? false,
    is_prohibited: opts.is_prohibited ?? false,
    effective_entity: opts.effective_entity ?? null,
    status_changes: dedupeStatusChanges(
      opts.statusIds.map((id) => getStatusChange(id))
    ),
    obligations: dedupeObligations(
      opts.obligationIds.map((id) => getObligation(id))
    ),
    determination_path: opts.path,
    answers: opts.answers,
    session_id: null,
  };
}

function entityObligations(
  entity: EntityType,
  isHighRisk: boolean
): string[] {
  const ids: string[] = [];
  if (entity === "provider") {
    if (isHighRisk) ids.push("provider_high_risk");
  } else if (entity === "deployer") {
    if (isHighRisk) ids.push("deployer_high_risk", "deployer_general");
    else ids.push("deployer_general");
  } else if (entity === "distributor") {
    if (isHighRisk) ids.push("distributor");
  } else if (entity === "importer") {
    if (isHighRisk) ids.push("importer");
  } else if (entity === "product_manufacturer") {
    ids.push("product_manufacturer");
    if (isHighRisk) ids.push("provider_high_risk");
  }
  return ids;
}

// ── main evaluate ─────────────────────────────────────────────────────────────

export function evaluate(answers: Record<string, unknown>): CheckerResult {
  const path: string[] = [];
  const statusIds: string[] = [];
  const obligationIds: string[] = [];

  // Gate: is this an AI system?
  if (!answerBool(answers, "gate_is_ai_system", true)) {
    path.push("gate:no");
    return buildResult({
      in_scope: false,
      path,
      statusIds: ["out_of_scope"],
      obligationIds: [],
      answers,
    });
  }
  path.push("gate:yes");

  const entity = answerEntity(answers);
  let effectiveEntity: EntityType = entity;
  path.push(`e1:${entity}`);

  // Authorised representative short-circuit
  if (entity === "authorised_rep") {
    return buildResult({
      in_scope: true,
      path,
      statusIds: [],
      obligationIds: ["authorised_representative"],
      effective_entity: effectiveEntity,
      answers,
    });
  }

  // Substantial modifications
  if (answerBool(answers, "e2_modifications")) {
    if (entity === "provider") {
      statusIds.push("handover");
      path.push("e2:handover");
    } else {
      statusIds.push("become_provider");
      effectiveEntity = "provider";
      path.push("e2:become_provider");
    }
  }

  // Product manufacturer with no integration → out of scope
  if (entity === "product_manufacturer") {
    const integration = answers["e3_product_integration"] ?? "none";
    if (integration === "none") {
      path.push("e3:none");
      return buildResult({
        in_scope: false,
        path,
        statusIds: dedupeIds([...statusIds, "out_of_scope"]),
        obligationIds: [],
        effective_entity: entity,
        answers,
      });
    }
  }

  const isHighRisk = determineHighRisk(answers);
  if (isHighRisk) {
    statusIds.push("high_risk");
    path.push("hr:high_risk");
  }

  // Geographic scope
  if (!answerBool(answers, "s1_in_scope", true)) {
    path.push("s1:none");
    return buildResult({
      in_scope: false,
      path,
      statusIds: dedupeIds([...statusIds, "out_of_scope"]),
      obligationIds: [],
      is_high_risk: isHighRisk,
      effective_entity: effectiveEntity,
      answers,
    });
  }
  path.push("s1:in_scope");

  // High-risk exception for EU deployers (Annex III only, no Annex I)
  const scopeRegion = answers["s1_scope_region"] ?? "eu";
  const hr1 = answerBool(answers, "hr1_annex_i");
  const hr2 = answerBool(answers, "hr2_annex_iii");
  if (
    effectiveEntity === "deployer" &&
    scopeRegion === "eu" &&
    hr2 &&
    !hr1 &&
    isHighRisk
  ) {
    statusIds.push("high_risk_exception");
    path.push("s1:high_risk_exception");
  }

  // Excluded
  if (answerBool(answers, "r2_excluded")) {
    path.push("r2:excluded");
    return buildResult({
      in_scope: false,
      path,
      statusIds: dedupeIds([...statusIds, "out_of_scope"]),
      obligationIds: [],
      is_high_risk: isHighRisk,
      effective_entity: effectiveEntity,
      answers,
    });
  }

  // Prohibited
  if (answerBool(answers, "r3_prohibited")) {
    path.push("r3:prohibited");
    return buildResult({
      in_scope: true,
      path,
      statusIds: dedupeIds([...statusIds, "prohibited"]),
      obligationIds: ["prohibited"],
      is_high_risk: isHighRisk,
      is_prohibited: true,
      effective_entity: effectiveEntity,
      answers,
    });
  }

  // AI literacy for provider + deployer
  if (effectiveEntity === "provider" || effectiveEntity === "deployer") {
    obligationIds.push("ai_literacy");
    path.push("e1:ai_literacy");
  }

  // GPAI
  if (answerBool(answers, "s1_gpai")) {
    statusIds.push("gpai");
    obligationIds.push("gpai_provider");
    path.push("r1:gpai");
    if (answerBool(answers, "s1_gpai_systemic_risk")) {
      obligationIds.push("gpai_systemic_risk");
      path.push("r1:gpai_systemic_risk");
    }
  }

  // Entity-role obligations
  obligationIds.push(...entityObligations(effectiveEntity, isHighRisk));

  // Transparency (limited-risk only)
  if (answerBool(answers, "r4_transparency") && !isHighRisk) {
    obligationIds.push("transparency");
    path.push("r4:transparency");
    const nonLiteracy = obligationIds.filter((id) => id !== "ai_literacy");
    if (
      (obligationIds.length === 2 &&
        obligationIds[0] === "ai_literacy" &&
        obligationIds[1] === "transparency") ||
      (nonLiteracy.length === 1 && nonLiteracy[0] === "transparency")
    ) {
      statusIds.push("transparency_only");
    }
  }

  // FRIA for deployers
  if (
    answerBool(answers, "r5_fria") &&
    isHighRisk &&
    effectiveEntity === "deployer"
  ) {
    obligationIds.push("fria");
    path.push("r5:fria");
  }

  return buildResult({
    in_scope: true,
    path,
    statusIds: dedupeIds(statusIds),
    obligationIds,
    is_high_risk: isHighRisk,
    effective_entity: effectiveEntity,
    answers,
  });
}

File: packages/core/src/opencomplai_core/compliance_checker/engine.py
"""Deterministic FLI-parity compliance checker engine."""

from __future__ import annotations

from typing import Any

from opencomplai_core.compliance_checker.catalog import (
    get_obligation,
    get_status_change,
)
from opencomplai_core.compliance_checker.models import (
    CHECKER_VERSION,
    CheckerSession,
    ComplianceCheckerResult,
    EntityType,
    ObligationItem,
    StatusChangeItem,
)

__all__ = ["CHECKER_VERSION", "evaluate"]


def _answer_bool(answers: dict[str, Any], key: str, default: bool = False) -> bool:
    value = answers.get(key, default)
    if isinstance(value, bool):
        return value
    if isinstance(value, str):
        return value.lower() in {"true", "yes", "1"}
    return bool(value)


def _answer_entity(answers: dict[str, Any]) -> EntityType:
    raw = answers.get("e1_entity_type", EntityType.PROVIDER.value)
    return EntityType(raw)


def _determine_high_risk(answers: dict[str, Any]) -> bool:
    hr1 = _answer_bool(answers, "hr1_annex_i")
    hr2 = _answer_bool(answers, "hr2_annex_iii")
    if not (hr1 or hr2):
        return False
    if _answer_bool(answers, "hr3_art_6_3"):
        return False
    if _answer_bool(answers, "hr4_narrow_task"):
        return False
    if _answer_bool(answers, "hr5_no_significant_risk"):
        return False
    if _answer_bool(answers, "hr6_accessory"):
        return False
    return True


def _dedupe_obligations(items: list[ObligationItem]) -> list[ObligationItem]:
    seen: set[str] = set()
    ordered: list[ObligationItem] = []
    for item in items:
        if item.id in seen:
            continue
        seen.add(item.id)
        ordered.append(item)
    return ordered


def _dedupe_status(items: list[StatusChangeItem]) -> list[StatusChangeItem]:
    seen: set[str] = set()
    ordered: list[StatusChangeItem] = []
    for item in items:
        if item.id in seen:
            continue
        seen.add(item.id)
        ordered.append(item)
    return ordered


def _build_result(
    *,
    in_scope: bool,
    path: list[str],
    status_ids: list[str],
    obligation_ids: list[str],
    is_high_risk: bool = False,
    is_prohibited: bool = False,
    effective_entity: EntityType | None = None,
    answers: dict[str, Any] | None = None,
) -> ComplianceCheckerResult:
    return ComplianceCheckerResult(
        checker_version=CHECKER_VERSION,
        in_scope=in_scope,
        determination_path=path,
        status_changes=[get_status_change(status_id) for status_id in status_ids],
        obligations=_dedupe_obligations(
            [get_obligation(obligation_id) for obligation_id in obligation_ids]
        ),
        is_high_risk=is_high_risk,
        is_prohibited=is_prohibited,
        effective_entity=effective_entity,
        answers=answers or {},
    )


def _entity_obligations(
    entity: EntityType,
    *,
    is_high_risk: bool,
) -> list[str]:
    obligation_ids: list[str] = []

    if entity == EntityType.PROVIDER:
        if is_high_risk:
            obligation_ids.append("provider_high_risk")
    elif entity == EntityType.DEPLOYER:
        if is_high_risk:
            obligation_ids.extend(["deployer_high_risk", "deployer_general"])
        else:
            obligation_ids.append("deployer_general")
    elif entity == EntityType.DISTRIBUTOR:
        if is_high_risk:
            obligation_ids.append("distributor")
    elif entity == EntityType.IMPORTER:
        if is_high_risk:
            obligation_ids.append("importer")
    elif entity == EntityType.PRODUCT_MANUFACTURER:
        obligation_ids.append("product_manufacturer")
        if is_high_risk:
            obligation_ids.append("provider_high_risk")

    return obligation_ids


def evaluate(session: CheckerSession) -> ComplianceCheckerResult:
    """Evaluate a checker session using FLI July 2025 flowchart logic."""
    answers = session.answers
    path: list[str] = []
    status_ids: list[str] = []
    obligation_ids: list[str] = []

    if not _answer_bool(answers, "gate_is_ai_system", default=True):
        path.append("gate:no")
        return _build_result(
            in_scope=False,
            path=path,
            status_ids=["out_of_scope"],
            obligation_ids=[],
            answers=answers,
        )
    path.append("gate:yes")

    entity = _answer_entity(answers)
    effective_entity = entity
    path.append(f"e1:{entity.value}")

    if entity == EntityType.AUTHORISED_REP:
        return _build_result(
            in_scope=True,
            path=path,
            status_ids=[],
            obligation_ids=["authorised_representative"],
            effective_entity=effective_entity,
            answers=answers,
        )

    if _answer_bool(answers, "e2_modifications"):
        if entity == EntityType.PROVIDER:
            status_ids.append("handover")
            path.append("e2:handover")
        else:
            status_ids.append("become_provider")
            effective_entity = EntityType.PROVIDER
            path.append("e2:become_provider")

    if entity == EntityType.PRODUCT_MANUFACTURER:
        integration = answers.get("e3_product_integration", "none")
        if integration == "none":
            path.append("e3:none")
            return _build_result(
                in_scope=False,
                path=path,
                status_ids=_dedupe_status_ids([*status_ids, "out_of_scope"]),
                obligation_ids=[],
                effective_entity=entity,
                answers=answers,
            )

    is_high_risk = _determine_high_risk(answers)
    if is_high_risk:
        status_ids.append("high_risk")
        path.append("hr:high_risk")

    if not _answer_bool(answers, "s1_in_scope", default=True):
        path.append("s1:none")
        return _build_result(
            in_scope=False,
            path=path,
            status_ids=_dedupe_status_ids([*status_ids, "out_of_scope"]),
            obligation_ids=[],
            is_high_risk=is_high_risk,
            effective_entity=effective_entity,
            answers=answers,
        )
    path.append("s1:in_scope")

    scope_region = answers.get("s1_scope_region", "eu")
    hr1 = _answer_bool(answers, "hr1_annex_i")
    hr2 = _answer_bool(answers, "hr2_annex_iii")
    if (
        effective_entity == EntityType.DEPLOYER
        and scope_region == "eu"
        and hr2
        and not hr1
        and is_high_risk
    ):
        status_ids.append("high_risk_exception")
        path.append("s1:high_risk_exception")

    if _answer_bool(answers, "r2_excluded"):
        path.append("r2:excluded")
        return _build_result(
            in_scope=False,
            path=path,
            status_ids=_dedupe_status_ids([*status_ids, "out_of_scope"]),
            obligation_ids=[],
            is_high_risk=is_high_risk,
            effective_entity=effective_entity,
            answers=answers,
        )

    if _answer_bool(answers, "r3_prohibited"):
        path.append("r3:prohibited")
        return _build_result(
            in_scope=True,
            path=path,
            status_ids=_dedupe_status_ids([*status_ids, "prohibited"]),
            obligation_ids=["prohibited"],
            is_high_risk=is_high_risk,
            is_prohibited=True,
            effective_entity=effective_entity,
            answers=answers,
        )

    if effective_entity in {EntityType.PROVIDER, EntityType.DEPLOYER}:
        obligation_ids.append("ai_literacy")
        path.append("e1:ai_literacy")

    if _answer_bool(answers, "s1_gpai"):
        status_ids.append("gpai")
        obligation_ids.append("gpai_provider")
        path.append("r1:gpai")
        if _answer_bool(answers, "s1_gpai_systemic_risk"):
            obligation_ids.append("gpai_systemic_risk")
            path.append("r1:gpai_systemic_risk")

    obligation_ids.extend(
        _entity_obligations(effective_entity, is_high_risk=is_high_risk)
    )

    if _answer_bool(answers, "r4_transparency") and not is_high_risk:
        obligation_ids.append("transparency")
        path.append("r4:transparency")
        non_literacy = [oid for oid in obligation_ids if oid != "ai_literacy"]
        if obligation_ids == ["ai_literacy", "transparency"] or (
            len(non_literacy) == 1 and non_literacy[0] == "transparency"
        ):
            status_ids.append("transparency_only")

    if (
        _answer_bool(answers, "r5_fria")
        and is_high_risk
        and effective_entity == EntityType.DEPLOYER
    ):
        obligation_ids.append("fria")
        path.append("r5:fria")

    return _build_result(
        in_scope=True,
        path=path,
        status_ids=_dedupe_status_ids(status_ids),
        obligation_ids=obligation_ids,
        is_high_risk=is_high_risk,
        effective_entity=effective_entity,
        answers=answers,
    )


def _dedupe_status_ids(status_ids: list[str]) -> list[str]:
    seen: set[str] = set()
    ordered: list[str] = []
    for status_id in status_ids:
        if status_id in seen:
            continue
        seen.add(status_id)
        ordered.append(status_id)
    return ordered
